### PR TITLE
Selectors reworked ...

### PR DESCRIPTION
**Pull request by cabo**
This is a full rebase of branch pr-98 (#99) on current main.

### CODE REVIEW

File: draft-ietf-jsonpath-base.md
Comment (cabo) [inline draft-ietf-jsonpath-base.md:674]: In the next PR, I'd prefer if we could move all the expression language parts over to an expression language section.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [inline draft-ietf-jsonpath-base.md:704]: See https://www.ietf.org/archive/id/draft-bormann-cbor-cddl-freezer-07.html#name-err6527 for how to do this right.  (Again, this should be in the expression language section.)

File: draft-ietf-jsonpath-base.md
Comment (cabo) [inline draft-ietf-jsonpath-base.md:1019]: What is the meaning of `<expr`?
Missing `>`?

Comment (goessner) [reply draft-ietf-jsonpath-base.md:1019]: yes ... simple typo

thanks a lot

**Review by goessner — APPROVED** [review]

File: draft-ietf-jsonpath-base.md
Comment (goessner) [reply draft-ietf-jsonpath-base.md:704]: Wouldn't make it sense then to simple copy that into here ?

**Comment by gregsdennis** [comment]
Does this incorporate any changes from the discussions from #98, or is it a simple rebase?

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [inline draft-ietf-jsonpath-base.md:640]: Having this as a parenthetical seems good to me. 👍

**Comment by cabo** [comment]
> Does this incorporate any changes from the discussions from #98, or is it a simple rebase?

It is a simple rebase, followed by a number of edits coming from the discussions here.
Most of them need some further discussion, so I hope to cover them with my updated slides in a couple of hours.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:716]: Yes.
(Discuss today.)

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [inline draft-ietf-jsonpath-base.md:799]: This doesn't quite match up with what's actually going on.  Hence [the issue](https://github.com/jsonpath-standard/jsonpath-compliance-test-suite/issues/3) in the compliance test suite I raised.

I was thinking it worked as you've written, but the referenced algorithm doesn't work exactly that way.

**Review by gregsdennis — CHANGES_REQUESTED** [review]
There are also still some open questions from #98.

File: draft-ietf-jsonpath-base.md
Comment (goessner) [reply draft-ietf-jsonpath-base.md:739]: Glyn already suggested to phrase instead: integer base-10

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:720]: Well, slight massaging required to support both single-quoted and double-quoted strings.
I think we should leave this to a followup PR.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:770]: penultimate

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [reply draft-ietf-jsonpath-base.md:739]: Right, but that's in point 2.  We still highlight "octal-like" integers in point 3.  Is there a reason we want to call attention to octal over other common number bases?

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [reply draft-ietf-jsonpath-base.md:692]: Suggested in meeting to use `string-literal`

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md]: Resolved as per interim (needed a bit more surgery than I had thought, though).

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md]: Resolved as per interim.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md]: Please check as now updated.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:778]: We didn't discuss this, so I left it, but "penultimate" really is the way to go.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:810]: We didn't discuss this, so I left it for a future PR

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md]: I didn't see this yet when I made a different change.  Please check...

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:775]:
```suggestion
For example, selector `[0]` selects the first and selector `[4]` the fifth element of a sufficiently long array.
```

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:795]: Same issue as before - don't put syntax in the semantics section.
```suggestion
An `index-wild-selector`
```

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:797]: Should talk about selecting nodes rather than values.

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:827]: @gregsdennis pointed out before that omitting the three integers always leaves behind _two_ separating colons when only one colon is valid. The ABNF is correct, but the description here contradicts it.

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:989]: The descendant selector should select all primitive values which are part of the subtree of the input node. For example, `..*` applied to the argument `true` should produce `[true]` even though there are no object members or array elements in the argument.

The descendant selector should also select _all_ object members and _all_ array elements (the current wording could permit some to be omitted).
```suggestion
It selects the node and all its descendants..
```

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1013]: For the record, "ws" is far more memorable than "S".

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1061]: We also need absolute path values such as `$.foo`.

**Review by glyn — CHANGES_REQUESTED** [review]
I left some comments or suggestions. I am now going on vacation, so apologies if I don't participate further for a couple of weeks.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:57]: I put him back to the contributors section (sorry about that).

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:166]: That's my German (nouns with caps, adjectives with lower case).  Fixed.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:604]: Indeed.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:613]: Changed this -- a bit redundant now.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:668]: Can't leave the "it" floating; slightly rephrased.

Comment (cabo) [reply draft-ietf-jsonpath-base.md:668]: Oh, you have better text below.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:796]: Fixed.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:826]: Wait, the step is optional in the syntax, too.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md]: To do: Define "descendants" (making sure that member values are, but member names aren't).

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1012]: If you are stuck in an ABNF littered with them, S is sufficiently memorable...

(WS really should be for mandatory whitespace, OWS is optional whitespace, but that is too long here.)

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1060]: Todo for next PR...

File: draft-ietf-jsonpath-base.md
Comment (sthagen) [inline draft-ietf-jsonpath-base.md:584]: I have a hard time understanding what a node of a value might be. This seems to be dimensionally wrong. Maybe there are too many objects in that sentence?

File: draft-ietf-jsonpath-base.md
Comment (sthagen) [reply draft-ietf-jsonpath-base.md:1025]: I never got that color thing people associate with space - maybe I should buy me a printer and do some archaeology? Just  `S` or `SP` should do nicely.

**Review by sthagen — APPROVED** [review]
Looks like a good next step on our path.

File: draft-ietf-jsonpath-base.md
Comment (glyn) [reply draft-ietf-jsonpath-base.md:839]: The ABNF syntax is correct (last time I looked!). It's just the above sentence which is misleading.

**Review by glyn — APPROVED** [review]
As I'm going away, approving so I don't block progress. I think @cabo has responded to most if not all my comments.

File: draft-ietf-jsonpath-base.md
Comment (goessner) [reply draft-ietf-jsonpath-base.md:1025]: I am fine with "S" (minimalistic :). Is there a convention to write constants in capitals?

File: draft-ietf-jsonpath-base.md
Comment (goessner) [reply draft-ietf-jsonpath-base.md:1073]: this is a `json-path` then.

File: draft-ietf-jsonpath-base.md
Comment (goessner) [inline draft-ietf-jsonpath-base.md:845]: Seems to be misleading (I know my wording). It was surely adopted by Python before 2014. So skip `later`.

File: draft-ietf-jsonpath-base.md
Comment (goessner) [inline draft-ietf-jsonpath-base.md:1067]: For the next PR:

Thinking a bit more about `in-op` leads me to the conclusion, that this operator makes sense only for arrays.
`'key' in $.obj` is equivalent to `$.obj.key` and `'key' in <literal-object>` seems to be useless. I see no other rational here. Someone else ?

**Review by goessner — APPROVED** [review]

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [reply draft-ietf-jsonpath-base.md:584]: I agree that this is rather verbose now.  Maybe

```suggestion
Finally, `.b` selects from any input node of type object any node corresponding to
a member named `b`.
```

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [reply draft-ietf-jsonpath-base.md]: You just added "As in JSON numbers..."

>3. As in JSON numbers, the syntax does not allow octal-like integers with leading zeros such as `01` or `-01`.

That doesn't address my concern at all.  My question is about the callout of _octal_ and exclusion of other number bases such as binary and hexadecimal.

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [inline draft-ietf-jsonpath-base.md:1001]: Repitition of "It" and extra period (ironically).

```suggestion
The `descendant-selector` is inspired by ECMAScript for XML (E4X).
It selects the node and all its descendants.
```

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [reply draft-ietf-jsonpath-base.md:1025]: Do we need to make concession for multiple whitespace chars (e.g. `*ws` or `*S`), or is that implicitly covered?

**Review by gregsdennis — APPROVED** [review]
Added some comments that I'm happy to have resolved in a follow-up non-functional edit PR.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1073]: That is a simple solution that we can refine later.  Put it in.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1025]: RFC 8610 defines this as:

     S = *WS
     WS = SP / NL
     SP = %x20
     NL = COMMENT / CRLF
     COMMENT = ";" *PCHAR CRLF
     PCHAR = %x20-7E / %x80-10FFFD
     CRLF = %x0A / %x0D.0A

I'm not sure we need to support comments in JSONPath (that is a serious question though), but if you remove that, pesky details like isolated CRs are fixed, too (and then CDDL simply disallows HTs, which may not be popular here).

Next PR...

Comment (cabo) [reply draft-ietf-jsonpath-base.md:1025]: Oh, and I'm not sure these can all be called constants, but there is indeed a tendency in ABNF usage to give token-level (scanner) productions such as DIGIT, ALPHA, WS, ... upper case names.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1067]: there is still the "enum" use case, as in `$.service_level in ["silver", "gold"]`, which is an abbreviation for `$.service_level == "silver" || $.service_level == "gold"`.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:845]: A quick search finds it in Python in 2009, but I didn't check further.
Changed to ", and that of Python".

BTW, do we support negative steps?

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1001]: Fixed.

File: draft-ietf-jsonpath-base.md
Comment (goessner) [reply draft-ietf-jsonpath-base.md:1069]: well ... this is the `in <array>` use case.

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [reply draft-ietf-jsonpath-base.md:846]: The current test suite and the comparison matrix _does_ have negative step cases.

Comment (glyn) [reply draft-ietf-jsonpath-base.md:846]: @cabo Negative steps are also supported in the spec. See for example Table 4.

Comment (cabo) [reply draft-ietf-jsonpath-base.md:846]: Thanks.  It just occurred to me as the semantics section needs to gain more content.

**Pull request merged**
